### PR TITLE
[Merged by Bors] - Added documentation on the query filters

### LINKED ISSUE #1551: Ambiguous system order and Changed<Component>
## Bevy version

source = "git+https://github.com/bevyengine/bevy?branch=main#d9fb61d474518eba06d0c3dce5cb81db9f7ed66a"

## Operating system & version

Linux

## What you did

Updated from bevy 0.4.0 to latest main and played my game

## What you expected to happen

That my systems would continue to run as before

## What actually happened

Many of my systems became unreliable - they weren't executing or were executing late (causing timeouts on a networking system). Different builds adding `dbg!()` commands changed behaviour etc. etc.

After some digging, I discovered this is likely due to the change: https://github.com/bevyengine/bevy/pull/1144

This change means that systems no longer execute in the order you declare them in. Coupled with the fact that the original Changed<component> issue explicitly states that order independance is a non goal: https://github.com/bevyengine/bevy/issues/54#issuecomment-662750622
This means systems with Changed appeared unreliable.

## Additional information

Technically this isn't a problem. The program is doing as documented: `Changed<component>` states it only works when systems are in order, AppBuilder explicitly states that systems aren't added in any sort of order. Both of those rules are fine. The issue is that the interaction between them is non-obvious to first-time users.

A minimum solution could be documentation on Changed<> noting this interaction and suggesting that any use of it requires explicit system order.

### PR DESCRIPTION
**Pull request by sdfgeoff**
This documents both the non-obvious interaction with non-explicit system ordering
and adds examples for Changed and Added. This likely closes #1551

### CODE REVIEW

**Comment by sdfgeoff** [comment]
Hmm, I'm not familiar enough with rustdoc to figure out how to make those examples pass CI. Can someone give me some pointers?

**Comment by Ratysz** [comment]
`cargo test --doc` in `bevy_ecs` directory.

File: crates/bevy_ecs/src/core/filter.rs [hunk 1]
```diff
@@ -31,12 +31,52 @@ pub struct Or<T>(pub T);
...
 /// Added components do not count as mutated.
+/// This filter is useful as a performance optimization as it means that a system can be made to run only when
+/// it needs to rather than every frame.
+///
+/// Because the ordering of systems can change and this filter is only effective on changes before the query executes
+/// you need to use explicit dependency ordering or ordered stages for these query filters to be useful.
+///
+/// Example:
+/// ```
+/// use bevy::prelude::*;
```
Comment (mockersf) [inline crates/bevy_ecs/src/core/filter.rs:42]: you can't use `use bevy::prelude::*;` from inside `bevy_ecs`, you will need to use a more direct path to the struct you want to use... and as `bevy_ecs` is one of the core crate, you don't have access to a lot of stuff without adding dev dependencies.

`Mutated`, `Added` and `Changed` should be reachable as they are defined here, but you can't use `Name` or `Transform`. Simplest solution would probably to create struct in the doc example and have them hidden, see https://doc.rust-lang.org/rustdoc/documentation-tests.html#hiding-portions-of-the-example

File: crates/bevy_ecs/src/core/filter.rs [hunk 2]
```diff
@@ -31,12 +31,52 @@ pub struct Or<T>(pub T);
...
+///
+/// fn print_moving_objects_system(query: Query<(Name,), (Mutated<Transform>,)>) {
```
Comment (mockersf) [inline crates/bevy_ecs/src/core/filter.rs:44]: no need to wrap the types in tuples, you can have
```rust
fn print_moving_objects_system(query: Query<Name, Mutated<Transform>>) {
    for name in query.iter() {
        println!("Entity Moved: {}", name)
    }
}
```

**Review by mockersf — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/core/filter.rs (hunk 2 above)
Comment (sdfgeoff) [reply crates/bevy_ecs/src/core/filter.rs]: Cool! I didn't know that, I've been putting all my queries in tuples.......

File: crates/bevy_ecs/src/core/filter.rs (hunk 1 above)
Comment (sdfgeoff) [reply crates/bevy_ecs/src/core/filter.rs]: Thanks for the review! Yup, I've got the examples compiling by using `bevy_ecs::` and hiding some structs.

**Comment by mockersf** [comment]
also, #1525 is changing the location of those comments, let's see what gets merged first 😄

**Comment by mockersf** [comment]
now defined [here](https://github.com/bevyengine/bevy/blob/3a2a68852c0a1298c0678a47adc59adebe259a6f/crates/bevy_ecs/src/query/filter.rs#L575)

**Comment by sdfgeoff** [comment]
Updated :)

File: crates/bevy_ecs/src/query/filter.rs [hunk 3]
```diff
@@ -574,6 +574,29 @@ macro_rules! impl_flag_filter {
 
 impl_flag_filter!(
     /// Filter that retrieves components of type `T` that have been added since the start of the frame
+    ///
+    /// This filter is useful as a performance optimization as it means that a system can be made to run only when
```
Comment (cart) [inline crates/bevy_ecs/src/query/filter.rs:578]: This isn't _quite_ true. The system will still run, but the query will only return changed items. Its a small distinction, but I think we should be crisp here.

Comment (sdfgeoff) [reply crates/bevy_ecs/src/query/filter.rs]: Indeed! I had to think about this one, but yep, system execution != query result. Thanks for pointing this out.

**Pull request closed without merging**